specification/dartLangSpec.tex

```diff
@@ -21253,39 +21261,2076 @@ \subsubsection{Informal Subtype Rule Descriptions}
+  \DefEquals{\UpperBoundType{$X_1$}{$T_2$}}{%
+    \left\{%
+    \begin{array}{l}
+      T_2\mbox{, if \SubtypeNE{X_1}{T_2}.}\\
+      X_1\mbox{, otherwise and if \SubtypeNE{T_2}{X_1}.}\\
+      \UpperBoundType{$B_{1a}$}{$T_2$}\mbox{,}\\
+      \mbox{\quad{}otherwise.}
+    \end{array}
+    \right.%
+  }\\[1mm]
+  where $B_{1a}$ is the greatest closure of $B_1$ with respect to $X_1$
+  (\ref{leastAndGreatestClosureOfTypes}).
+\item
+  Let $X_1$ be a type variable with bound $B_1$,
+  and $S_1$ a subtype of $B_1$.\\[1mm]
+  \DefEquals{\UpperBoundType{$X_1$\,\&\,$S_1$}{$T_2$}}{%
+    \left\{%
+    \begin{array}{l}
+      %% TODO(eernst), for review: Why do we not have the following instead?
+      % T_2\mbox{, if \SubtypeNE{\code{$X_1$\,\&\,$S_1$}}{T_2}.}\\
+      T_2\mbox{, if \SubtypeNE{X_1}{T_2}.}\\
+      %% TODO(eernst), for review: Why do we not have the following as well?
+      % X_1 \& S_1\mbox{, otherwise and if \SubtypeNE{T_2}{X_1 \& S_1}.}\\
+      X_1\mbox{, otherwise and if \SubtypeNE{T_2}{X_1}.}\\
+      \UpperBoundType{$S_{1a}$}{$T_2$}\mbox{,}\\
+      \mbox{\quad{}otherwise.}
+    \end{array}
+    \right.%
+  }\\[1mm]
+  where $S_{1a}$ is the greatest closure of $S_1$ with respect to $X_1$.
+\item
+  Let $X_2$ be a type variable with bound $B_2$.\\[1mm]
+  \DefEquals{\UpperBoundType{$T_1$}{$X_2$}}{%
+    \left\{%
+    \begin{array}{l}
+      X_2\mbox{, if \SubtypeNE{T_1}{X_2}.}\\
+      T_1\mbox{, otherwise and if \SubtypeNE{X_2}{T_1}.}\\
+      \UpperBoundType{$T_1$}{$B_{2a}$}\mbox{,}\\
+      \mbox{\quad{}otherwise.}
+    \end{array}
+    \right.%
+  }\\[1mm]
+  where $B_{2a}$ is the greatest closure of $B_2$ with respect to $X_2$.
+\item
+  %% TODO(eernst), for review: Consider the same changes as in the converse.
+  Let $X_2$ be a type variable with bound $B_2$,
+  and $S_2$ a subtype of $B_2$.\\[1mm]
+  \DefEquals{\UpperBoundType{$T_1$}{$X_2$\,\&\,$S_2$}}{%
+    \left\{%
+    \begin{array}{l}
+      X_2\mbox{, if \SubtypeNE{T_1}{X_2}.}\\
+      T_1\mbox{, otherwise and if \SubtypeNE{X_2}{T_1}.}\\
+      \UpperBoundType{$T_1$}{$B_{2a}$}\mbox{,}\\
+      \mbox{\quad{}otherwise.}
+    \end{array}
+    \right.%
+  }\\[1mm]
+  where $S_{2a}$ is the greatest closure of $S_2$ with respect to $X_2$.
+\item
+  \DefEquals{\UpperBoundType{$T$ \FUNCTION<\ldots>(\ldots)}{\FUNCTION}}{%
+    \FUNCTION}.
+\item
+  \DefEquals{\UpperBoundType{\FUNCTION}{$T$ \FUNCTION<\ldots>(\ldots)}}{%
+    \FUNCTION}.
+\item
+  Let $U_1$ and $U_2$ be, respectively,
+
+  \noindent
+  \code{$T_1$\,\FUNCTION<$X_1$\,\EXTENDS\,$B_{11}$,\,\ldots,\,$X_m$\,%
+    \EXTENDS\,$B_{1m}$>($P_{11}$,\,\ldots,\,$P_{1k}$)}
+
+  \noindent
+  \code{$T_2$\,\FUNCTION<$X_1$\,\EXTENDS\,$B_{21}$,\,\ldots,\,$X_m$\,%
+    \EXTENDS\,$B_{2m}$>($P_{21}$,\,\ldots,\,$P_{2l}$)}
+
+  \noindent
+  such that each $B_{1i}$ and $B_{2i}$ are types with the same canonical syntax,
+  and both have the same number of required positional parameters.
+  Let $q$ be $\metavar{min}(k, l)$,
```

Comment:
Interesting. I think both the CFE and the Analyzer expect `k == l`. At least they both accept the code below that shows that `UP(Function(int, int, int), Function(int, int)) = Function` and not `Function(int, int)`, as this rule suggests.

```dart
foo(void Function(int, int, int) f, void Function(int, int) g) {
  var x = [f, g];
  var h = x.first;
  h(0, 1, 2, 3, 4, 5);
}
```